cmd, console: split off the console into a reusable package

karalabe opened the pull request:
Our console handling became quite messy, with a lot of JavaScript black magic being scattered all over out repository. This PR aims to clean all of that up and move the console to its own package where it can be developed without mucking around with the geth main command, also allowing it to be tested at a package level, without doing the heavyweight geth node startup and teardown.

The PR is mostly a huge cleanup of everything that pertains to the console. Feature wise it enhances the console to support custom input and output streams, allowing proper testing (i.e. without needing to run a dedicated process to is to access stdin/stdout); but also allowing the reuse of the console in projects that embed geth into their own processes.

Lastly it moves the our jsre package into an internal namespace. That package is quite a horrid thing, mostly hacking ottovm to work as required by our console and provides zero usecase to anything but our console (also it's ugly as hell by its very nature).

Trophies: 

 * WebSocket attach failed on OSX: https://github.com/ethereum/go-ethereum/pull/2539
 * Available module listing was missing on console/attach
 * `rpc.DefaultIPCApis` was not sorted alphabetically (nitpick)
 * The RPC layer had a hidden special namespace (expose it): https://github.com/ethereum/go-ethereum/pull/2548
 * One of our shh web3 extensions wasn't actually exposed: https://github.com/ethereum/go-ethereum/pull/2548
 * `miner.setGasPrice` had an invalid input formatter assigned: https://github.com/ethereum/go-ethereum/pull/2548
 * The geth integration tester was using HTML templates, not text ones: https://github.com/ethereum/go-ethereum/pull/2549
 * Some of our methods threw exceptions while others printer error messages as plain text.
 * Change the history access to `0600` instead of `0777`.

---

The second addon commit colorizes the JavaScript exceptions too. Until now they were just dumped as is, without any colors.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue May 10 10:09:08 UTC 2016_

fjl commented on cmd/geth/consolecmd_test.go:
```diff
@@ -15,3 +15,139 @@
+	blob, _ := ioutil.ReadAll(geth.stdout)
+	output := string(blob)
+
+	if want := "Welcome"; !strings.Contains(output, want) {
```
you can use `geth.expect("template code")` here

karalabe replied:
Can it work around text colorization?

fjl replied:
No, but the console shouldn't output ANSI color escapes unless the output is a terminal.

fjl commented on cmd/geth/consolecmd_test.go:
```diff
@@ -15,3 +15,138 @@
+ datadir: {{datadir}}{{end}}
+ modules:{{range apis}} {{.}}:1.0{{end}}
+
+{{prompt}}
```
Not sure it is much more readable with so many functions. It might be easier to just
put the actual output text in the template. For example, the coinbase is the same in all tests.

So instead of defining the template helper (twice!), adapting the tester to track it, you can just put
it in as text:

```text
coinbase: 0x8605cdbbdb6d264aa742e77020dcbc58fcdce182
at block: 0 (Thu, 01 Jan 1970 00:00:00 UTC)
```

(I know that it outputs the time in the local zone right now but that should probably change as well).

karalabe replied:
Not sure about that tbh. Imho it provides more information to the user about the current status of the chain than having a random timezone you need to calculate with.

fjl commented on cmd/geth/consolecmd_test.go:
```diff
@@ -15,3 +15,138 @@
+	attach.setTemplateFunc("etherbase", func() string { return geth.Etherbase })
+	attach.setTemplateFunc("niltime", func() string { return time.Unix(0, 0).Format(time.RFC1123) })
+	attach.setTemplateFunc("ipc", func() bool { return strings.HasPrefix(endpoint, "ipc") })
+	attach.setTemplateFunc("datadir", func() string { return geth.Datadir })
```
Note that public fields on `geth` are available in the template. You don't need to define this wrapper, just use `{{.Datadir}}`.

karalabe replied:
I need `geth.Datadir`, whereas the template is run against `attach`. I.e. I have to geth instances here.

fjl replied:
fair point

karalabe commented:
@fjl @bas-vk Please review this PR. It's more or less done. I'm thinking about adding a few more tests to check password prompts and other custom injected JavaScript methods, but there shouldn't be any significant changes.

fjl commented on console/prompter.go:
```diff
@@ -0,0 +1,156 @@
+// TerminalPrompter holds the stdin line reader (also using stdout for printing
+// prompts). Only this reader may be used for input because it keeps an internal
+// buffer.
+var TerminalPrompter = newTerminalPrompter()
```
Please call this one `Stdin`.

karalabe replied:
Fixed, will push in next commit.

fjl commented:
LGTM apart from one comment. Please rebase.

bas-vk commented:
The current console allows a user to stop the console with `exit`. This is not implemented in the new console. I think this needs to be supported.
```
> exit
ReferenceError: 'exit' is not defined
    at <anonymous>:1:1
```
It would better if there was a something like `console.exit(<optional message>)` which can be called from javascript. If optional message is given the process prints the message and exists with an error code.

bas-vk commented:
The history functionality has changed. If you type in a command and press the "up" key it won't find the previous line. Also it overwrites the history file where the old console would append to an existing history file.

fjl commented:
Maybe I should actually test this before saying LGTM.

bas-vk commented on console/console.go:
```diff
@@ -0,0 +1,365 @@
+			}
+			// If all the needed lines are present, save the command and run
+			if indents <= 0 {
+				if len(input) == 0 || passwordRegexp.MatchString(input) || !leadingSpace.MatchString(input) {
```
I have the feeling a not is missing.

karalabe replied:
Yeah, this is something i asked felix too. The old console had this same bug

karalabe replied:
Fixed (pulled in @fjl 's changes from a previous bugfix pr).

karalabe replied:
I've also fixed `exit`. It was a small refactor error, I used the wrong variable to check for the exit string. Fixed on next commit.

karalabe commented:
I'll rebase this when @bas-vk two PRs are merged in. He made a few changes that conflict with this and I don't want to do the messy rebase twice.

karalabe commented:
@fjl @bas-vk PTAL

I've addressed all the raised issued. Please do check that history works, passwords are not saved, etc. I've also changed the history file premissions to `0600`.

bas-vk commented on console/console.go:
```diff
@@ -0,0 +1,370 @@
+
+// Stop cleans up the console and terminates the runtime envorinment.
+func (c *Console) Stop(graceful bool) error {
+	if err := ioutil.WriteFile(c.histPath, []byte(strings.Join(c.history, "\n")), 0600); err != nil {
```
In case the file already exists the permissions are not set.
Maybe do `os.Chmod`.

karalabe replied:
Fair point. I thought it would be enough to have it for new files, but you're right that we should set the correct permissions for old ones as well. Fixed and amended.

bas-vk commented:
Apart from that small comment it looks 👍.
I will take a stab at https://github.com/ethereum/go-ethereum/issues/2443 when this is merged.

fjl commented on console/prompter.go:
```diff
@@ -0,0 +1,165 @@
+
+	// SetScrollHistory sets the the input scrollback history that the prompter will
+	// allow the user to scoll back to.
+	SetScrollHistory(history []string)
```
It feels weird to call this "scroll history". Why can't it just be `SetHistory`, `AppendHistory`?
This should not prevent merge but please feel free to change these names if you agree.

karalabe replied:
Fair enough. It seemed a bit clearer, given that `liner`calls it `scrollback history`, but I'll change it. There's really no need to make it this explicit.

karalabe replied:
Fixed.

fjl commented:
👍
